crypto/evp/evp_pkey.c

```diff
@@ -73,6 +74,12 @@ EVP_PKEY *EVP_PKCS82PKEY_ex(const PKCS8_PRIV_KEY_INFO *p8, OSSL_LIB_CTX *libctx,
     int selection;
     size_t len;
     OSSL_DECODER_CTX *dctx = NULL;
+    const ASN1_OBJECT *algoid = NULL;
+    char keytype[OSSL_MAX_NAME_SIZE];
+
+    if (!PKCS8_pkey_get0(&algoid, NULL, NULL, NULL, p8)
+            || !OBJ_obj2txt(keytype, sizeof(keytype), algoid, 0))
+        return NULL;
```

Comment:
Shouldn't we iterate decoders in this case?

Reply:
It should be falling back in that case I would imagine..

Reply:
So we should not return NULL here, right?

Reply:
Better might be `keytype = NULL` instead of `return NULL`.  Needing `keytype` to be a pointer not an array which introduces some mess.

Reply:
> Better might be `keytype = NULL` instead of `return NULL`. Needing `keytype` to be a pointer not an array which introduces some mess.

Just introduce a second variable `char *pkeytype = &keytype`, and pass that.

Reply:
> Shouldn't we iterate decoders in this case?

Why? Neither of these calls should ever fail in the normal course of events. `OBJ_obj2txt()` will return the OID in textual form if it is something unrecognised. If they do fail then some fundamental error has occurred.

Reply:
Then we will attempt to match the decoder by OID text string (which all of our own decoders supply). This was the suggestion made by by @levitte in https://github.com/openssl/openssl/pull/20412#discussion_r1121769993 and is the result of using `OBJ_obj2txt` instead of `OBJ_nid2sn` as in the original patch.

Basically we have two choices. Either we use `OBJ_obj2txt` which cannot fail in normal operation, and will either give us the name of the object if it is known, or the OID in textual form if it is unknown. Or we can use the `OBJ_obj2nid(OBJ_nid2sn())` approach (as per the original patch) which will give us the name if it is known, or nothing if it is unknown.

If we use `OBJ_obj2txt` then we *always* have a name of some form except in the case of fundamental error, and so we don't fallback to using a NULL name ever. If we use `OBJ_obj2nid(OBJ_nid2sn())` then we may get a NULL name and we should fallback to the full decoding.

Reply:
So if 3rd-party decoder doesn't match an OID string, we consider it as a bug in the decoder, right?

If so, it's worth documenting it.

Reply:
@beldmit, do you mean a provider that has an OSSL_ALGORITHM item like this:

``` C
    { "blargh", NULL, &blarch_functions },
```

... and then libcrypto gets to handle a PKCS#8 blob with the OID, say, 1.3.6.1.4.1.5168.4711.9999, which we can pretend is the OID for blargh.  How will that get to the correct provider, I guess the question is, right?

There are two ways to handle this:

1.  Add the OID in the name string, i.e. something like this:

    ``` C
        { "blargh:1.3.6.1.4.1.5168.4711.9999", NULL, &blarch_functions },
    ```

    This is the preferred method.

2.  Register the OID in libcrypto's OBJ db, using the upcall number `OSSL_FUNC_CORE_OBJ_CREATE`

    This is less preferred, 'cause it's frankly a lot of extra churn, and it involves dealing with NIDs, something we're trying to get away from.

Reply:
@levitte, right, thanks for confirmation! Then I'd say it needs to be documented (at least the 1st way).

Reply:
Hehe, parallel answering at work here ;-)

Yes, BTW, I would consider it a bug if a provider of algorithms libcrypto knows nothing about fails to provide the OID in some form.  It's not only about PKCS#8, I think that CMS and other similar systems/protocols have similar underlying issues, we just haven't really dealt with them yet.

Reply:
(just for emphasis: of course it's OK if some algorithm doesn't have an associated OID at all...  it can still be used in all places where an OID isn't expected, and will obviously never appear in structures and objects where an OID is expected)

Reply:
@beldmit, you might want to read this manual:

https://github.com/openssl/openssl/blob/master/doc/man3/OSSL_ALGORITHM.pod

Do you mean to say that it should be worded more strongly, or would you rather have a note on this in another manual?  Which one, in that case?

Reply:
@levitte If I understand correctly, before these changes are merged, we didn't have a mandatory requirement for OID to be present - and now we have it. So adding smth like "OIDs are mandatory in case they are not well-known (present in OID database of OpenSSL) to parse the private keys" looks reasonable here and probably in "provider-decoder" page.

Reply:
Hmmm. Actually this seems like a change in behaviour, so I'm now wondering whether the original `OBJ_obj2nid(OBJ_nid2sn())` was the correct answer after all...

Reply:
> Hmmm. Actually this seems like a change in behaviour, so I'm now wondering whether the original `OBJ_obj2nid(OBJ_nid2sn())` was the correct answer after all...

I really don't think so.  However, it's arguable that if `EVP_DECODE_CTX_new_from_pkey()` with the keytype parameter != NULL fails, we could fallback to a second call with the keytype parameter = NULL, on the assumption that the failure of the first call being that it couldn't find a keymgr for the given keytype.

Reply:
> @levitte If I understand correctly, before these changes are merged, we didn't have a mandatory requirement for OID to be present - and now we have it. So adding smth like `OIDs are mandatory in case they are not well-known (present in OID database of OpenSSL) to parse the private keys` looks reasonable here and probably in "provider-decoder" page.

You forget that this depends on the usage of the algorithm.  If all someone wants to do is `openssl enc -provider blargh -blargh -e < foo.txt`, there's no need for an OID.  So the forcefulness of the message needs to be tempered with regard to usage.

So what I said earlier about this being a bug in the provider was a bit too forceful.  I should have said that *for this usage*, I consider it a bug

Reply:
> I consider it a bug

Hmm. But, before this change, you could have a decoder that didn't specify the OID in textual form and didn't register the OID anywhere else, and `EVP_PKCS82PKEY_ex` would eventually find and use the decoder (after having tried a long list of other ones first). After this change the decoder will no longer be found.

Reply:
> > I consider it a bug
> 
> Hmm. But, before this change, you could have a decoder that didn't specify the OID in textual form and didn't register the OID anywhere else, and `EVP_PKCS82PKEY_ex` would eventually find and use the decoder (after having tried a long list of other ones first). After this change the decoder will no longer be found.

Hmmm, on re-reading [`OSSL_ALGORITHM(3)`](https://github.com/openssl/openssl/blob/master/doc/man3/OSSL_ALGORITHM.pod), I'd say it does suggest that for things that involve OIDs, the OID should be present in some form.  It doesn't say out right that things will absolutely fail, but it *is* recommended to, and this is a direct quote:

> make sure that the corresponding **OSSL_ALGORITHM**'s *algorithm_names* include known names as well as the OID itself in canonical decimal text form.  That should cover all scenarios.

To me, this suggests that if used to work with `EVP_PKCS8PKEY_ex()`, it was more by luck than anything else.

But granted, that's a harsh interpretation...

So what do you think about my suggestion of a fallback `EVP_DECODER_CTX_new_from_pkey` call with keytype = NULL?

Reply:
I'm sorry, I forgot...  the way to detect if no decoders were found is with `OSSL_DECODER_CTX_get_num_decoders()`.  example code that displays this:

``` C
#include <stdio.h>
#include <openssl/decoder.h>
#include <openssl/evp.h>
#include <openssl/err.h>

int main()
{
  EVP_PKEY *key = NULL;
  OSSL_DECODER_CTX *ctx;

  if ((ctx = OSSL_DECODER_CTX_new_for_pkey(&key, NULL, NULL, "blargh",
                                           0, NULL, NULL))) {
    printf("Number of decoders found for 'blargh': %d\n",
           OSSL_DECODER_CTX_get_num_decoders(ctx));
  } else {
    ERR_print_errors_fp(stderr);
  }
  OSSL_DECODER_CTX_free(ctx);
}
```

``` console
$ cc -o foo foo.c -lcrypto
$ ./foo
Number of decoders found for 'blargh': 0
```

Reply:
> So what I said earlier about this being a bug in the provider was a bit too forceful. I should have said that _for this usage_, I consider it a bug

Smth like:
```
All the names necessary for using an algorithm (symbolic name for looking up by name, OID for usage in ASN.1 structures etc) should be explicitly provided in case they are not well-known (e.g. present in OID database of OpenSSL)
```

Reply:
> > So what I said earlier about this being a bug in the provider was a bit too forceful. I should have said that _for this usage_, I consider it a bug
> 
> Smth like: `All the names necessary for using an algorithm (symbolic name for looking up by name, OID for usage in ASN.1 structures etc) should be explicitly provided in case they are not well-known (e.g. present in OID database of OpenSSL)`

PR welcome ;-)

Reply:
Sure :)

Reply:
IMO it is actually wrong to go ahead with the decoding with keytype=NULL. It could be misused to parse incorrect PKCS8 structures where the OID does not match the actual contents of the private key and the decoder would decode it just by chance. Better to fail in that case as currently in this PR. So IMO with the docfix as suggested by @beldmit I think the current PR is good as is.

Reply:
> It could be misused to parse incorrect PKCS8 structures where the OID does not match the actual contents of the private key and the decoder would decode it just by chance. 

Well that would be a clear bug in the decoder. Conversely failing to register the OID or declare the OID text as an alias is not a *clear* bug (there is at least some debate about whether it is a bug or not) - and we could break existing decoders by requiring it. This would put a big question mark to me over whether or not this is suitable as a bug fix for 3.0, or even whether we would want to introduce it at this late stage for 3.1. Whereas the original version of this PR I would have no such concerns over.

> I'm sorry, I forgot... the way to detect if no decoders were found is with OSSL_DECODER_CTX_get_num_decoders(). example code that displays this:

So your proposal, IIUC, is to keep the logic as it currently is using `OBJ_obj2txt`, but if after the `OSSL_DECODER_CTX_new_for_pkey` we find that `OSSL_DECODER_CTX_get_num_decoders()` is 0, then re-call `OSSL_DECODER_CTX_new_for_pkey` but with a NULL keytype? This sounds plausible.

Reply:
> So your proposal, IIUC, is to keep the logic as it currently is using `OBJ_obj2txt`, but if after the `OSSL_DECODER_CTX_new_for_pkey` we find that `OSSL_DECODER_CTX_get_num_decoders()` is 0, then re-call `OSSL_DECODER_CTX_new_for_pkey` but with a NULL keytype? This sounds plausible.

Exactly correct.

Reply:
See #20422